nulib/meadow
# Extract digest generation from pipeline to S3 (or Minio) event

# Summary 


The checksum generation was taking too long to complete (1 hour +) for super large video files (50G and up) as a step in our ingest pipeline. We decided to extract checksum generation from the pipeline and use AWS's [Serverless Fixity for Digital Preservation Compliance](https://docs.aws.amazon.com/solutions/latest/serverless-fixity-for-digital-preservation-compliance/welcome.html) step function triggered by an upload to the uploads or ingest buckets. For now, we are generating an `md5` checksum (staging/prod)


# Specific Changes in this PR

- Adds auto-checksum functionality to ingest and upload minio buckets in dev and test environments
- Update the `meadow.buckets.create` task to configure bucket notifications
- Add configuration info to `dev.exs` and `test.exs` to support the above 
- Terraform for Cloud Trail and Event Bridge to invoke the step function
- Refactor `GenerateFileSetDigests` pipeline action to use the checksums stored in the object tags rather than calculate them
- Use the Elixir `WaitForIt` package where appropriate to wait for checksums to be generated before kicking off ingest
- Update GraphQL API to return map of all checksums
- Update Preservation tab's checksum copy functionality to accommodate new API
- Update `Validator` to check for object tags and report error (we consider this to be the easiest first iteration - not necessarily the best long term solution) [Discussion issue for future](https://github.com/nulib/repodev_planning_and_docs/issues/2469)
- Update preservation check with `md5` column
- __NOTE: for now - 1st iteration - the step function itself has been manually installed__

# Version bump required by the PR

See [Semantic Versioning 2.0.0](https://semver.org/) for help discerning which is required.

- [x] Major (backwards incompatible GraphQL API change)

# Steps to Test

1. Set up the environment
    ```
    devstack down
    devstack update
    devstack up -d meadow
    mix meadow.buckets.create
    ```
2. Start your dev environment server
3. Run an ingest sheet ingest
4. Verify that checksum hash is present in the GraphQL API
5. Verify that the copy checksum functionality on the preservation tab in the UI 
6. Perform a UI file uplopad
7. Verify that the copy checksum functionality on the preservation tab in the UI 
7. Verify that checksum hash is present in the GraphQL API
3. Launch the Minio console in a web browser
3. Upload an object to the `dev-ingest` (or `dev-uploads`) bucket
4. Click on the newly uploaded object to see its details
5. Make sure there are `computed-[hash]` and `computed-[hash]-last-modified` tags for each hash type (`md5` and `sha1`). If the uploaded object it large, it might take several seconds for the tags to show up.
6. Make sure the auto-tagging does _not_ happen on other buckets.
7. You can `devstack logs -f minio-checksum` if you want to see the log output from the hash-tagging webhook
8. Let the preservation check run, download it and observe the md5 column

# :rocket: Deployment Notes

**Note** - __if you check any of these boxes go to the [(always open) `main` <- `staging` PR](https://github.com/nulib/meadow/pulls) and add detailed notes and instructions to help out others who may end up deploying your changes to production__

- Backward compatible API changes
- Backwards-incompatible API changes
  - [x] Adds/requires new or changed Terraform variables [`fixity_function`](https://github.com/nulib/tfvars/commit/9a6ece5f8d6dabc7f7063f775d58dd38c5c0d021) - we still need prod added to `tfvars`
  -  **__Do not merge until after both #2682 and nulib/devstack#9 have been merged__**
  - requires all developers to run `devstack update` and restart the stack before using meadow again
  - Manually install and configure [Serverless Fixity for Digital Preservation Compliance](https://docs.aws.amazon.com/solutions/latest/serverless-fixity-for-digital-preservation-compliance/welcome.html) on production


# Tested/Verified
